bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,531 @@
+                    String gsonDateStr = gson.toJson(new Date());
+                    _storage.put(LAST_USED.getKey(handle), gsonDateStr);
+
+                    if (!allHandles.contains(handle)) {
+                        // update all handles index
+                        allHandles.add(handle);
+                        _storage.put(STORE_KEY_INDEX_OF_HANDLES, gson.toJson(allHandles));
+                    }
+                }
+            } finally {
+                storageLock.unlock();
+            }
+        }
+
+        public void put(String handle, PersistedParams persistedParams) {
+            try {
+                storageLock.lock();
+                if (_storage == null) {
+                    throw new RuntimeException(EXCEPTION_STORAGE);
+                }
+
+                if (persistedParams == null) {
+                    _storage.put(SERVICE_CONFIGURATION.getKey(handle), (String) null);
+                } else {
+                    String gsonPersistedParamsStr = gson.toJson(persistedParams);
+                    _storage.put(SERVICE_CONFIGURATION.getKey(handle), gsonPersistedParamsStr);
+                    String gsonDateStr = gson.toJson(new Date());
+                    _storage.put(LAST_USED.getKey(handle), gsonDateStr);
+                    if (!allHandles.contains(handle)) {
+                        // update all handles index
+                        allHandles.add(handle);
+                        _storage.put(STORE_KEY_INDEX_OF_HANDLES, gson.toJson(allHandles));
+                    }
+                }
+            } finally {
+                storageLock.unlock();
+            }
+        }
+
+        public void removeByHandle(String handle) {
+            try {
+                logger.info("Removing handle {} from storage", handle);
+                storageLock.lock();
+                if (_storage == null) {
+                    // just log the error and let it be, do not throw exception
+                    logger.warn(EXCEPTION_STORAGE);
+                } else {
+                    _storage.remove(ACCESS_TOKEN_RESPONSE.getKey(handle));
+                    _storage.remove(LAST_USED.getKey(handle));
+                    _storage.remove(SERVICE_CONFIGURATION.getKey(handle));
+                }
+            } finally {
+                storageLock.unlock();
+            }
+        }
+
+        public void removeAll() {
+            // no need any locks, the other methods will take care of this
+            HashSet<String> allHandles = getAllHandlesFromIndex();
+            for (String handle : allHandles) {
+                removeByHandle(handle);
+            }
+        }
+
+        @Override
+        public void close() {
+            boolean lockGained = false;
+            try {
+                // dont want to wait too long during shutdown or update
+                lockGained = storageLock.tryLock(15, TimeUnit.SECONDS);
```

Comment:
lock should be outside of the try block

Reply:
if it throws InterruptedException, it would try to unlock the monitor, even though it is not active anymore.

Reply:
lockGained would be false if it was not successful.  The close-without-lock is therefore taken care of.  tryLock itself may throw InterruptedException which I intend to catch.